### PR TITLE
Added query condition support

### PR DESCRIPTION
**Pull request by DimitrisStaratzis**

### CODE REVIEW

File: src/main/java/io/tiledb/spark/Range.java [hunk 1]
```diff
@@ -96,43 +102,43 @@ public Range merge(Range other) {
       Pair<Byte, Byte> otherRange = other.range;
       return new Range(
           new Pair(
-              min(range.getFirst(), otherRange.getFirst()),
-              max(range.getSecond(), otherRange.getSecond())));
+              max(range.getFirst(), otherRange.getFirst()),
```
Comment (DimitrisStaratzis) [inline src/main/java/io/tiledb/spark/Range.java:105]: We need to get the intersection, not union.

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java [hunk 2]
```diff
@@ -318,15 +326,49 @@ private boolean initQuery() throws TileDBError {
     query = new Query(array, QueryType.TILEDB_READ);
 
     // Pushdown any ranges
-    if (pushedRanges.size() > 0) {
-      for (List<Range> ranges : pushedRanges) {
+    QueryCondition finalCondition = null;
+    if (allRanges.size() > 0) {
+      int j = 0;
+      // the first element of the allranges list is a list of the dimension ranges. The remaining
```
Comment (ihnorton) [inline src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:332]: This logic could be simplified with (IIUC) no overhead by taking two sublists -- the first element and the trailing elements -- and handle the dimension and QC cases in separate `for` loops.

File: src/main/java/io/tiledb/spark/TileDBDataSourceReader.java [hunk 3]
```diff
@@ -446,62 +482,73 @@ public boolean enableBatchRead() {
...
+      ranges.get(columnIndex).add(new Range(new Pair<>(f.value(), f.value())));
 
       // GreaterThan is ranges which are in the form of `dim > 1`
     } else if (filter instanceof GreaterThan) {
       GreaterThan f = (GreaterThan) filter;
-      int dimIndex = this.tileDBReadSchema.getDimensionId(f.attribute()).get();
+      Object second;
+      if (nonEmptyDomain.get(f.attribute()) != null)
+        second = nonEmptyDomain.get(f.attribute()).getSecond();
+      else second = getMaxValue(f.value().getClass());
```
Comment (ihnorton) [inline src/main/java/io/tiledb/spark/TileDBDataSourceReader.java:498]: style nit: newline and indent `second`

File: src/main/java/io/tiledb/spark/Range.java [hunk 4]
```diff
@@ -18,7 +18,13 @@
 
   Range(Pair range) {
     this.range = range;
-    this.dataClassType = range.getFirst().getClass();
+    if (range.getFirst() != null) this.dataClassType = range.getFirst().getClass();
+    else if (range.getSecond() != null) this.dataClassType = range.getSecond().getClass();
```
Comment (ihnorton) [inline src/main/java/io/tiledb/spark/Range.java:22]: Missing `else` handling?

File: src/main/java/io/tiledb/spark/Range.java [hunk 5]
```diff
@@ -18,7 +18,13 @@
...
+  }
+
+  Range(boolean empty, Pair range) {
+    this.range = range;
+    this.dataClassType = null;
```
Comment (ihnorton) [inline src/main/java/io/tiledb/spark/Range.java:27]: Is this valid? I don't think I see any checks of `dataClassType` where it uses `null`? What is supposed to happen in that case?

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java [hunk 6]
```diff
@@ -4,6 +4,9 @@
 import static io.tiledb.java.api.QueryStatus.TILEDB_COMPLETED;
 import static io.tiledb.java.api.QueryStatus.TILEDB_INCOMPLETE;
 import static io.tiledb.java.api.QueryStatus.TILEDB_UNINITIALIZED;
+import static io.tiledb.libtiledb.tiledb_query_condition_combination_op_t.TILEDB_AND;
```
Comment (ihnorton) [inline src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:7]: nit: this should be enum under `io.tiledb.java.api.QueryCondition`, right?

(ok to leave for follow-up though)

File: src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java [hunk 7]
```diff
@@ -234,6 +251,117 @@ public void testHeterogeneousSparse1() throws TileDBError {
...
+      int[] d2 = new int[] {12, 40, 50};
+      int[] a1 = new int[] {10, 23, 30};
+      int[] a2 = new int[] {100, 230, 300};
+
+      for (int i = 0; i < rows1.size(); i++) {
+        Assert.assertEquals(d1[i], rows1.get(i).getString(0));
+        Assert.assertEquals(d2[i], rows1.get(i).getInt(1));
+        Assert.assertEquals(a1[i], rows1.get(i).getInt(2));
+        Assert.assertEquals(a2[i], rows1.get(i).getInt(3));
+        //        System.out.println(
```
[the remark this answers is not in the document]
Comment (DimitrisStaratzis) [reply src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java]: Done!

File: src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java [hunk 8]
```diff
@@ -234,6 +251,117 @@ public void testHeterogeneousSparse1() throws TileDBError {
...
+              .read()
+              .format("io.tiledb.spark")
+              .option("uri", SPARSE_ARRAY_URI)
+              .option("order", order)
+              .option("partition_count", 1)
+              .load();
+      dfRead.createOrReplaceTempView("tmp");
+      dfRead.show();
+      List<Row> rows1 = dfRead.sqlContext().sql("SELECT * FROM tmp").collectAsList();
+      List<Row> rows2 = dfRead.sqlContext().sql("SELECT * FROM tmp WHERE a1 > 25").collectAsList();
```
[the remark this answers is not in the document]
Comment (DimitrisStaratzis) [reply src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java]: I added comments next to each condition. The or condition is not pushed down and uses the built-in filtering.

File: src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java [hunk 9]
```diff
@@ -234,6 +251,117 @@ public void testHeterogeneousSparse1() throws TileDBError {
...
+      List<Row> rows3 = dfRead.sqlContext().sql("SELECT * FROM tmp WHERE a1 >= 23").collectAsList();
+      List<Row> rows4 =
+          dfRead.sqlContext().sql("SELECT * FROM tmp WHERE a1 > 25 or a1 < 20").collectAsList();
+      List<Row> rows5 =
+          dfRead.sqlContext().sql("SELECT * FROM tmp WHERE a1 > 25 AND a2 > 250").collectAsList();
```
[the remark this answers is not in the document]
Comment (DimitrisStaratzis) [reply src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java]: Done!

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java (hunk 2 above)
Comment (DimitrisStaratzis) [reply src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:331]: Done

File: src/main/java/io/tiledb/spark/TileDBDataSourceReader.java (hunk 3 above)
Comment (DimitrisStaratzis) [reply src/main/java/io/tiledb/spark/TileDBDataSourceReader.java:498]: All formatting is done by ./gradlew format.

File: src/main/java/io/tiledb/spark/Range.java (hunk 4 above)
Comment (DimitrisStaratzis) [reply src/main/java/io/tiledb/spark/Range.java:22]: No need to handle that case. The classType will be null.

File: src/main/java/io/tiledb/spark/Range.java (hunk 5 above)
Comment (DimitrisStaratzis) [reply src/main/java/io/tiledb/spark/Range.java]: Removed it, it was unnecessary.

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java (hunk 6 above)
Comment (DimitrisStaratzis) [reply src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:7]: Not sure what you mean

File: src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ihnorton) [reply src/test/java/io/tiledb/spark/TileDBDataSourceReadTest.java]: Thanks

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java (hunk 2 above)
Comment (ihnorton) [reply src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:331]: Much easier to read now, IMO -- thank you!

File: src/main/java/io/tiledb/spark/Range.java (hunk 4 above)
Comment (ihnorton) [reply src/main/java/io/tiledb/spark/Range.java:22]: Thanks

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java (hunk 6 above)
Comment (ihnorton) [reply src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:7]: I mean that the other enums are translated from `tiledb_*_*_t` to something like `io.tiledb.java.api.QueryStatus.TILEDB_UNINITIALIZED` or `QueryType.TILEDB_READ` so we should do the same here for consistency -- doesn't need to be in this PR though.

File: src/main/java/io/tiledb/spark/TileDBDataSourceReader.java (hunk 3 above)
Comment (ihnorton) [reply src/main/java/io/tiledb/spark/TileDBDataSourceReader.java:498]: Ok, we can blame the robot 😄

**Review by ihnorton — APPROVED** [review]
LGTM, thanks for addressing the comments! Let's get this merged! :shipit: 
If @Shelnutt2 has any feedback we can do follow-up PRs to address later.

File: src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java (hunk 6 above)
Comment (ihnorton) [reply src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java:7]: Nevermind, looks like `QueryStatus` is the only one like this.

**Pull request merged**